src/core/importDataSources.ts

```diff
@@ -0,0 +1,190 @@
+import { URL } from 'whatwg-url';
+import { getFileMimeType } from '../io';
+import { readRemoteManifestFile } from '../io/manifest';
+import { ARCHIVE_FILE_TYPES } from '../io/mimeTypes';
+import { extractFilesFromZip } from '../io/zip';
+import Pipeline, { Handler } from './pipeline';
+import { canFetchUrl, fetchFile } from '../utils/fetch';
+
+/**
+ * Represents a URI source with a file name for the downloaded resource.
+ */
+export interface UriSource {
+  uri: string;
+  name: string;
+}
+
+/**
+ * Represents a user-specified file.
+ */
+export interface FileSource {
+  file: File;
+  fileType: string;
+}
+
+/**
+ * If an archive source is specified, then it is assumed that the data source
+ * has a FileSource (representing the file inside the archive), and a parent
+ * data source that has a FileSource of the containing archive.
+ */
+export interface ArchiveSource {
+  // Full path + filename inside the archive
+  path: string;
+}
+
+/**
+ * Cases to handle:
+ *   associated dataset file is part of a (potentially nested) remote archive
+ */
+export interface DataSource {
+  uriSrc?: UriSource;
+  fileSrc?: FileSource;
+  archiveSrc?: ArchiveSource;
+  parent?: DataSource;
+}
+
+function isArchive(r: DataSource): r is DataSource & { fileSrc: FileSource } {
+  return !!r.fileSrc && ARCHIVE_FILE_TYPES.has(r.fileSrc.fileType);
+}
+
+/**
+ * Transforms a file data source to have a mime type
+ * @param dataSource
+ */
+const retypeFile: Handler<DataSource> = async (dataSource, { next }) => {
+  let src = dataSource;
+  const { fileSrc } = src;
+  if (fileSrc && fileSrc.fileType === '') {
+    const mime = await getFileMimeType(fileSrc.file);
+    if (mime) {
+      src = {
+        ...src,
+        fileSrc: {
+          ...fileSrc,
+          fileType: mime,
+        },
+      };
+    }
+  }
+  next(src);
+};
+
+const importSingleFile: Handler<DataSource> = (dataSource, { next }) => {
+  if (dataSource.fileSrc) {
+    // pass to readers
+    console.log('importing single file', dataSource);
+  } else {
+    next();
+  }
+};
+
+const importDicomFile: Handler<DataSource> = (dataSource, { next }) => {
+  if (dataSource.fileSrc?.fileType === 'application/dicom') {
+    // pass to dicom store
+    console.log('importing dicom', dataSource);
+  } else {
+    next();
+  }
+};
+
+/**
+ * Extracts files from an archive
+ * @param dataSource
+ */
+const extractArchive: Handler<DataSource> = async (
+  dataSource,
+  { next, execute }
+) => {
+  if (isArchive(dataSource)) {
+    const files = await extractFilesFromZip(dataSource.fileSrc.file);
+    files.forEach((entry) => {
+      execute({
+        fileSrc: {
+          file: entry.file,
+          fileType: '',
+        },
+        archiveSrc: {
+          path: entry.archivePath,
+        },
+        parent: dataSource,
+      });
+    });
+  } else {
+    next();
```

Comment:
Should `next` just be the returned value?  Then normal functions can be middleware and encourage `return` for control flow rather than logic nesting.

Would that mean we need "sink" middleware to call a `endExecution()` func to avoid pushing undefined around?  And promise resolving in the middleware glue?

Reply:
That should be doable. If a handler were to return rather than call next, then we'd need to detect undefined, which is fine. Nobody should be transforming data to undefined.
- Returning undefined means terminate the pipeline
- else, continue with that input

We would have to disable the eslint rule that requires consistent returns in all code paths, otherwise we'd have odd-looking `return undefined` everywhere.

```
const handler = (input, { execute }) => {
  if (isArchive(input)) {
    execute(extract(input))
    // can also disable eslint consistent returns
    return undefined
  }
  return input
}
```

Related, I am playing with having pipelines optionally return data as part of the pipeline result. This would mean that handlers must call `done(outputData)` if they want to terminate and produce data. Do you suppose it would make more sense for a non-undefined return value to designate the return value of a pipeline, or a transformed output of the handler?

Reply:
Here is a simpler approach. We will keep your idea of having handlers return the (optionally transformed, optionally wrapped in a Promise) input, and change the usage of `done()`. Calling `done()` signals to the pipeline that, when the handler returns, the pipeline should be terminated and the return value should be discarded.

```
const handler = (input, { execute, done }) => {
  if (isArchive(input)) {
    execute(await extract(input))
    // terminate processing of input
    done()
  }
  return input
}
```

Reply:
Explicit `done()` makes sense.  

Just musing: another semantic option which makes sense for this particular unwrapping, recursive use case: `execute()` is always done with returned values.  But opting into recursion makes more sense in the general case probably.

Reply:
I would rather have `execute()` run with explicit arguments, since it's more obvious. Middleware return values are strictly to pass along to the next pipeline handler. It's possible that a handler wants to execute a nested pipeline _and_ pass control along to the next handler.